packages/flutter/lib/src/rendering/object.dart

```diff
@@ -1803,28 +1887,53 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   @protected
   bool get alwaysNeedsCompositing => false;
 
-  OffsetLayer _layer;
   /// The compositing layer that this render object uses to repaint.
   ///
-  /// Call only when [isRepaintBoundary] is true and the render object has
-  /// already painted.
-  ///
-  /// To access the layer in debug code, even when it might be inappropriate to
-  /// access it (e.g. because it is dirty), consider [debugLayer].
-  OffsetLayer get layer {
-    assert(isRepaintBoundary, 'You can only access RenderObject.layer for render objects that are repaint boundaries.');
-    assert(!_needsPaint);
+  /// If this render object is not a repaint boundary, it is the responsibility
+  /// of the [paint] method to populate this field. If [needsCompositing] is
+  /// true, this field may be populated with the root-most layer used by the
+  /// render object implementation. When repainting, instead of creating a new
+  /// layer the render object may update the layer stored in this field for better
+  /// performance. It is also OK to leave this field as null and create a new
+  /// layer on every repaint, but without the performance benefit. If
+  /// [needsCompositing] is false, this field must be set to null either by
+  /// never populating this field, or by setting it to null when the value of
+  /// [needsCompositing] changes from true to false.
+  ///
+  /// If this render object is a repaint boundary, the framework automatically
+  /// creates an [OffsetLayer] and populates this field prior to calling the
+  /// [paint] method. The [paint] method must not replace the value of this
+  /// field.
+  ContainerLayer get layer {
+    assert(!isRepaintBoundary || (_layer == null || _layer is OffsetLayer));
     return _layer;
   }
+
+  // Protect the setter because it only makes sense for [RenderObject]
+  // implementations to set the layer.
+  @protected
+  set layer(ContainerLayer newLayer) {
+    assert(
+      !isRepaintBoundary,
+      'Attempted to set a layer to a repaint boundary render object.\n'
+      'The framework creates and assigns an OffsetLayer to a repaint '
+      'boundary automatically.',
+    );
+    _layer = newLayer;
+  }
+  ContainerLayer _layer;
+
   /// In debug mode, the compositing layer that this render object uses to repaint.
   ///
   /// This getter is intended for debugging purposes only. In release builds, it
   /// always returns null. In debug builds, it returns the layer even if the layer
   /// is dirty.
   ///
   /// For production code, consider [layer].
-  OffsetLayer get debugLayer {
-    OffsetLayer result;
+  // Deprecated as of July 17, 2019.
```

Comment:
why do we want to remove this?

Reply:
Because any render object can create and keep a layer the normal `layer` getter no longer requires `isRepaintBoundary` and `!_needsPaint`. So this getter no longer adds any value to tests. They can as successfully use the `layer` getter (in fact, I moved all our existing tests to it).

Reply:
It seems to me that this should still require `needsPaint`. All the more so now than before. Indeed the setter probably should have asserts too: we presumably should never change our layer except during paint.

Reply:
In the past it required `!_needsPaint`:

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/rendering/object.dart#L1816

With this change it no longer requires it, because, as you say we're producing layers during paint. Are you suggesting that we should flip the requirement to require `_needsPaint`?

Reply:
Actually, we're clearing `_needsPaint` before calling `paint`, so requiring `_needsPaint` in the `layer` getter won't work:

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/rendering/object.dart#L2107

Reply:
I meant that `layer` should require `!needsPaint`, sorry for the confusion above.

If we make `layer` protected I guess I'm not as worried about this. Though it sounds like we still want the `debugLayer` for those places outside of RenderObject that want to read the layer during development.